Add ssz_types crate

paulhauner opened the pull request:
## Issue Addressed

Resolves #431

## Proposed Changes

Add a new `ssz_types` crate that unifies the following structs:

- `FixedVector`: a fixed-length list. Implementation is very much copied from the existing `FixedLenVec`.
- [NEW] `VariableList`: a variable-length list with a maximum length.
- [NEW] `BitVector`: fixed-capacity bitfield.
- [NEW] `BitList`: variable length bitfield with a maximum length.

## Additional Info

This PR _should_ implement SSZ serialization as per `v0.8.0`, however it is not checked against EF test vectors.

This PR _does not_ provide the following:

- As mentioned, it is not tested against EF test vectors.
- Implement tree hashing (or cached tree hashing) as per the spec. This will be done as a part of #432.
 - Replace the existing `FixedLenVec` with the new `FixedVector`. This should be done, but is left for another PR.
- Implement any of the new structs in the rest of the codebase. This should be done, but is left for another PR; @michaelsproul will likely want to be involved in this. It is tracked in #433.
- Make the crate compatible with wasm targets.

paulhauner commented:
@michaelsproul this is ready for review and use in your dev branches :)

I'm going to set my sights on #432

michaelsproul commented on eth2/utils/ssz_types/src/bitfield.rs:
```diff
@@ -0,0 +1,1114 @@
+    where
+        D: Deserializer<'de>,
+    {
+        // We reverse the bit-order so that the BitVec library can read its 0th
```
This comment is out of date now that we're not using `BitVec`

michaelsproul commented on eth2/utils/ssz_types/src/fixed_vector.rs:
```diff
@@ -0,0 +1,335 @@
+}
+
+impl<T, N: Unsigned> FixedVector<T, N> {
+    /// Returns `Some` if the given `vec` equals the fixed length of `Self`. Otherwise returns
```
Returns `Ok`/`Err` rather than `Some`/`None`

michaelsproul commented on eth2/utils/ssz_types/src/bitfield.rs:
```diff
@@ -0,0 +1,1114 @@
+///
+/// See the [`Bitfield`](struct.Bitfield.html) docs for usage.
+#[derive(Clone, PartialEq, Debug)]
+pub struct BitList<N> {
```
What if we named these marker structs `Variable<N>` and `Fixed<N>`, and then had type aliases for `BitList` and `BitVector` like:

```rust
type BitList<N> = Bitfield<Variable<N>>;
type BitVector<N> = Bitfield<Fixed<N>>;
```

michaelsproul replied:
(I have done this via dodgy renaming in `lib.rs` on my branch, and I like it)

paulhauner replied:
Yeah I like this!

FYI, I experimented with only exposing the type alias, but if we don't expose `Bitfield` we don't get documentation for the `Bitfield` methods (which is basically all of them).

michaelsproul replied:
Cool, I'm happy for `Bitfield` and `Variable` and `Fixed` to all be exposed

paulhauner replied:
We can get away with not exposing `Variable` and `Fixed`, which is appealing to me.

michaelsproul replied:
Happy with that too

michaelsproul commented on eth2/utils/ssz_types/src/bitfield.rs:
```diff
@@ -0,0 +1,1114 @@
+    where
+        D: Deserializer<'de>,
+    {
+        // We reverse the bit-order so that the BitVec library can read its 0th
```
BitVec again

michaelsproul commented on eth2/utils/ssz_types/src/bitfield.rs:
```diff
@@ -0,0 +1,1114 @@
+
+            Some(())
+        } else {
+            None
```
I quite liked it when these functions (`get`/`set`) returned `Result`, because then it was straight-forward to convert the errors in `state_processing`.

michaelsproul replied:
i.e.

```rust
state.justification_bits.set(0, true)?;
```

rather than

```
        state
            .justification_bits
            .set(0, true)
            .ok_or(ssz_types::Error::OutOfBounds {
                i: 1,
                len: state.justification_bits.len(),
            })?;
```

paulhauner replied:
Yeah I preferred `Result` vs `Option` as well -- I just thought it was a little off compared to `Vec.get(..)`, etc.

If we both prefer `Result`, then lets do that. #listentoyourheart

michaelsproul replied:
Good point about `Vec`, but I think we have a strong case for the `Result`, and it's always easier to throw information away (`.ok()`) than to conjure it from thin air

michaelsproul commented on eth2/utils/ssz_types/src/bitfield.rs:
```diff
@@ -0,0 +1,1114 @@
+
+    /// Returns true if no bits are set.
+    pub fn is_zero(&self) -> bool {
+        !self.bytes.iter().any(|byte| (*byte & u8::max_value()) > 0)
```
Seems like a strange way to write this, the `& u8::max_value()` is a no-op, and `!any` is the same as `all!`:

```rust
self.bytes.iter().all(|byte| *byte == 0)
```

paulhauner commented:
@michaelsproul thanks for your detailed review, much appreciated! I believe I have addressed all your comments.

I did not move the bitfield bitwise ops (e.g., intersection) from `Option` to `Result` as I understand you're working on these elsewhere.
